core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -289,15 +294,14 @@ public void registerSuggester(String key, Suggester<?> suggester) {
     /**
      * Register a new ScoreFunctionParser.
      */
-    public void registerFunctionScoreParser(ScoreFunctionParser<? extends ScoreFunctionBuilder> parser) {
+    public <T extends ScoreFunctionBuilder<T>> void registerScoreFunction(ScoreFunctionParser<T> parser, Writeable.Reader<T> reader) {
         for (String name: parser.getNames()) {
-            Object oldValue = functionScoreParsers.putIfAbsent(name, parser);
+            Object oldValue = scoreFunctionParsers.putIfAbsent(name, parser);
             if (oldValue != null) {
                 throw new IllegalArgumentException("Function score parser [" + oldValue + "] already registered for name [" + name + "]");
             }
         }
-        @SuppressWarnings("unchecked") NamedWriteable<? extends ScoreFunctionBuilder> sfb = parser.getBuilderPrototype();
-        namedWriteableRegistry.registerPrototype(ScoreFunctionBuilder.class, sfb);
+        namedWriteableRegistry.register(ScoreFunctionBuilder.class, parser.getNames()[0], reader);
```

Comment:
If we don't want to address this as part of this PR, let's add some TODO or norelease. I think we should do the same that we do for queries: make the parser a functional interface and use ParseField for parsing.